src/MicrosoftTeams.ts

```diff
@@ -564,23 +736,64 @@ namespace microsoftTeams
             return;
         }
 
-        // Process only if the message is coming from a valid origin or the origin used for local testing
+        // Process only if the message is coming from a valid origin
+        let messageSource = evt.source || evt.originalEvent.source;
         let messageOrigin = evt.origin || evt.originalEvent.origin;
-        if (validOrigins.indexOf(messageOrigin.toLowerCase()) === -1)
+        if ((validOrigins.indexOf(messageOrigin.toLowerCase()) === -1) &&
+            (messageOrigin !== currentWindow.location.origin) ||
+            (messageSource === currentWindow))
         {
             return;
         }
 
-        // Set our parent origin so that we can use it when sending messages
-        parentOrigin = messageOrigin;
+        // Update our parent and child relationships based on this message
+        updateRelationships(messageSource, messageOrigin);
 
-        // If we have any messages in our queue send them now
-        while (messageQueue.length > 0)
+        // Handle the message
+        if (messageSource === parentWindow)
+        {
+            handleParentMessage(evt);
+        }
+        else if (messageSource === childWindow)
         {
-            parentWindow.postMessage(messageQueue.shift(), parentOrigin);
+            handleChildMessage(evt);
         }
+    }
 
-        // Check to see if this looks like a request or a response
+    function updateRelationships(messageSource: Window, messageOrigin: string): void
+    {
+        // Determine whether the source of the message is our parent or child and update our
+        // window and origin pointer accordingly
+        if (!parentWindow || (messageSource === parentWindow))
+        {
+            parentWindow = messageSource;
+            parentOrigin = messageOrigin;
+        }
+        else if (!childWindow || (messageSource === childWindow))
+        {
+            childWindow = messageSource;
+            childOrigin = messageOrigin;
+        }
+
```

Comment:
As discussed offline, we only call updateRelationships if the messageOrigin matches one of our validDomains or is equal to our current origin (i.e. window.location.origin). This is enforced by processMessage. Therefore, even if the parent or child get compromised and navigate to some different non-valid domain then we will ignore their messages.